iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -160,6 +160,23 @@ void VariableOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
   results.insert<InlineConstVariableOpInitializer>(context);
 }
 
+OpFoldResult VariableLoadOp::fold(ArrayRef<Attribute> operands) {
```

Comment:
I don't think we should add this. For example, if two routines use the same variable, then it can get inlined into both (and e.g. potentially quantized with different quantization settings) doubling binary size. Doing this certainly needs a cost model and goes well beyond what I feel comfortable in a fold.

Reply:
It's always correct to inline an immutable value as it cannot change - not by quantization or any other work - and if it's expected to be able to change then it shouldn't be marked immutable. Variables can be created mutable and then made immutable when there are no remaining writes after passes that may mutate them as you say are performed (of which there are none today).

Happy to improve this with cost modeling whenever we get to that point - for now, this leaves hundreds of variables of scalar constants in many cases. Constants are generally a mess in MLIR/XLA right now (only splats are folded in most cases, constants grow very large due to folding on broadcasts of splat values, etc), and I'd rather a systematic look at the whole problem of constant storage and global constant deduplication be taken instead of randomly hoping we don't fold in certain cases :)

Reply:
I think Sean's point is that by making it a fold you're saying "we always want to do this", whereas if you make it a separate pattern it opens you up to better cost models and such. It's the same reason that iota expansion shouldn't be a fold.

Reply:
I am indeed saying we always want to do this: flow.variable is not designed to act as some kind of implicit mechanism for forcing constant deduplication. It's meant to model stateful variable lifetime and that's it. If we want something that does constant-specific stuff we can add it (like flow.constant, or flow.constant_pool, etc). iota is different because it is actually synthesizing something new and pessimizes everything.